src/components/Pagination/Pagination.utilities.ts

```diff
@@ -4,70 +4,45 @@ export interface Page {
 }
 
 export const generatePages = (
-  pageRange: number,
   pageTotal: number,
   activePage: number,
   numberOfPagesDisplayed: number
 ): Page[] => {
+  const pageRange = Math.min(pageTotal, Math.max(1, numberOfPagesDisplayed));
   const pages: Page[] = [];
   let startingPage = 1;
   let endingPage = pageRange;
 
-  if (pageTotal <= pageRange) {
-    startingPage = 1;
-    endingPage = pageRange;
-  } else if (activePage + numberOfPagesDisplayed > pageTotal) {
-    startingPage = pageTotal - (numberOfPagesDisplayed - 1);
-    endingPage = startingPage + (numberOfPagesDisplayed - 1);
-  } else if (
-    activePage > numberOfPagesDisplayed &&
-    activePage + numberOfPagesDisplayed <= pageTotal
-  ) {
-    startingPage = activePage - Math.floor(numberOfPagesDisplayed / 2);
-    endingPage = startingPage + (numberOfPagesDisplayed - 1);
+  if (activePage + Math.floor(pageRange / 2) >= pageTotal) {
+    startingPage = Math.max(1, pageTotal - pageRange + 1);
+    endingPage = pageTotal;
+  } else {
+    startingPage = Math.max(1, activePage - Math.floor(pageRange / 2));
+    endingPage = Math.min(pageTotal, startingPage + pageRange - 1);
   }
 
   for (let i = startingPage; i <= endingPage; i += 1) {
     pages.push({ pageNumber: i, isPage: true });
   }
 
-  if (pageTotal > pages[pages.length - 1]?.pageNumber) {
-    const secondToLastPage =
-      pageTotal !== activePage + numberOfPagesDisplayed
-        ? activePage + numberOfPagesDisplayed
-        : pageTotal - 1;
-
-    // only add ellipsis if there are more than 0 pages between the final page and the rest of the pages
-    if (pageTotal > numberOfPagesDisplayed + 1) {
-      pages.push({ pageNumber: secondToLastPage, isPage: false });
+  // Handling ellipsis for overflow pages
+  if (endingPage < pageTotal) {
+    if (endingPage < pageTotal - 1) {
+      pages.push({ pageNumber: -1, isPage: false }); // represents ellipsis
     }
-
     pages.push({ pageNumber: pageTotal, isPage: true });
   }
 
-  if (activePage > numberOfPagesDisplayed) {
-    const threeDotsPage =
-      activePage - numberOfPagesDisplayed > 1
-        ? activePage - numberOfPagesDisplayed
-        : activePage - numberOfPagesDisplayed + 1;
-
-    pages.unshift(
-      { pageNumber: 1, isPage: true },
-      { pageNumber: threeDotsPage, isPage: false }
-    );
+  if (startingPage > 1) {
+    pages.unshift({ pageNumber: 1, isPage: true });
+    if (startingPage > 2) {
+      pages.splice(1, 0, { pageNumber: -1, isPage: false }); // represents ellipsis
+    }
   }
 
-  return [...pages];
+  return pages;
 };
```

Comment:
General refactor to:
* Remove unnecesary parameter `pageRange`, to calculate it inside the function.
* Simplify and improve the calculation of pages.
* Simplify and improve the calculation of ellipses.
* Remove complex condition like `if-else`.
* Use `Math` library to improve calcs.